src/to_unicode.cpp

```diff
@@ -0,0 +1,50 @@
+#include "ada/idna/to_unicode.h"
+
+#include "ada/idna/punycode.h"
+#include "ada/idna/unicode_transcoding.h"
+
+namespace ada::idna {
+
+std::string to_unicode(const std::string_view& input) {
+  std::string out{};
+  out.reserve(input.size());
+
+  const char* label_start = input.data();
+  const char* input_end = input.data() + input.size();
+
+  while (label_start < input_end) {
+    const char* loc_dot = std::find(label_start, input_end, '.');
+    bool is_last_label = (loc_dot == input_end);
+    size_t label_size =
+        is_last_label ? (input_end - label_start) : (loc_dot - label_start);
+
+    const char* label_end = label_start + label_size;
+
+    if (label_size == 0) {
+      // empty label? Nothing to do.
+    } else if (ada::idna::verify_punycode(
```

Comment:
verify_punycode only checks that the punycode can be decoded, but it does not check that what is decoded is valid. (Is it mapped, has it been normalized?)